src/hardware/pcspeaker.cpp

```diff
@@ -18,493 +19,570 @@
-					value += spkr.volcur * vol_len;
-					const auto speed_by_len = spkr_speed * vol_len;
-					const auto speed_by_len_sq = speed_by_len *
-					                             vol_len / 2.0;
-					if (vol_diff < 0) {
-						value -= speed_by_len_sq;
-						spkr.volcur -= speed_by_len;
-					} else {
-						value += speed_by_len_sq;
-						spkr.volcur += speed_by_len;
-					}
-					index+=vol_len;
-				}
-			}
-		}
-		spkr.prev_pos = pos;
-		*stream++=(int16_t)(value/sample_add);
+	// Write silence if the waveform deque ran out
+	while (requested_frames) {
+		spkr.chan->AddSamples_m16(1, &SPKR_NEUTRAL_AMPLITUDE);
+		++spkr.tally_of_silence;
+		--requested_frames;
 	}
 
-	int16_t *buffer = reinterpret_cast<int16_t *>(MixTemp);
-	if (spkr.neutralize_dc_offset)
-		PlayOrFadeout(pos, len, buffer);
-	else
-		spkr.chan->AddSamples_m16(len, buffer);
+	// Maybe put the channel to sleep
+	if (spkr.tally_of_silence > 1000) {
+		spkr.pit_prev_amplitude = SPKR_NEUTRAL_AMPLITUDE;
+		spkr.chan->Enable(false);
+	}
+}
+
+static void InitializeImpulseLUT()
+{
+	assert(spkr.impulse_lut.size() == SPKR_FILTER_WIDTH);
+	for (uint16_t i = 0; i < SPKR_FILTER_WIDTH; ++i)
+		spkr.impulse_lut[i] = impulse(
+		        i / (static_cast<double>(spkr.rate) * SPKR_OVERSAMPLING));
 }
+
 class PCSPEAKER final : public Module_base {
 public:
 	PCSPEAKER(Section *configuration) : Module_base(configuration)
 	{
-		Section_prop * section=static_cast<Section_prop *>(configuration);
+		Section_prop *section = static_cast<Section_prop *>(configuration);
 		if (!section->Get_bool("pcspeaker"))
 			return;
-		spkr.rate = std::max(section->Get_int("pcrate"), 8000);
 
-		std::string dc_offset_pref = section->Get_string("zero_offset");
-		if (dc_offset_pref == "auto")
-#if !defined(WIN32)
-			spkr.neutralize_dc_offset = true;
-#else
-			spkr.neutralize_dc_offset = false;
-#endif
-		else
-			spkr.neutralize_dc_offset = (dc_offset_pref == "true");
-
-		spkr.dc_silencer.Configure(static_cast<uint32_t>(spkr.rate),
-		                           DC_SILENCER_WAVES, DC_SILENCER_WAVE_HZ);
-
-		spkr.min_tr = (PIT_TICK_RATE + spkr.rate / 2 - 1) / (spkr.rate / 2);
-		/* Register the sound channel */
-		spkr.chan = MIXER_AddChannel(&PCSPEAKER_CallBack, spkr.rate, "SPKR");
-		spkr.chan->SetPeakAmplitude(
-		        static_cast<uint32_t>(AMPLITUDE_POSITIVE));
+		// Get the playback rate and calculate related constants
+		spkr.rate = std::max(section->Get_int("pcrate"), 8000);
+		spkr.rate_as_float = static_cast<float>(spkr.rate);
+		spkr.rate_per_ms = spkr.rate_as_float / 1000.0f;
+		spkr.minimum_counter = 2 * PIT_TICK_RATE / spkr.rate;
+
+		InitializeImpulseLUT();
+
+		// Size the waveform queue
+		// +1 to compensate for rounding down of the division
+		static auto waveform_size = SPKR_FILTER_QUALITY +
+		                            spkr.rate / 1000 + 1;
+		spkr.waveform_deque.resize(check_cast<uint16_t>(waveform_size), 0.0f);
+
+		// Setup the soft limiter
+		constexpr auto channel_name = "SPKR";
```

Comment:
I was thinking of renaming `SPKR` to `PCSPEAKER` later, maybe you could do it now?
We use `INNOVATION` and not `INVTN`, and while we have `GUS` and `SB` and `FM`, those are well-known and reasonable abbreviations (although I'd prefer `OPL` over `FM`). `SPEAKER` would be another option, but it's too generic and can be misinterpreted when looking at the mixer output; `PCSPEAKER` is unambiguous.

Reply:
Definitely agree. The channel name become part of the public API though, via `mixer.com <CHANNEL_NAME> ...`, so there will be frontends and users with batch files that are relying on `SPKR`.

So for this, I'll add a silent fallback to map `SPKR` to `PCSPEAKER`, and also generate a log warning that `SPKR` is deprecated (and that it will eventually be removed, say it in a couple more versions).

Leaving this unresolved until this is in.

Reply:
Added a hidden alias + deprecation warning in https://github.com/dosbox-staging/dosbox-staging/pull/1690/commits/54329fb98e8739497bb7616b5a3d564378d23062:

```
2022-06-10 21:49:41.650 | MIXER: 'SPKR' is deprecated due to inconsistent
naming, please use 'PCSPEAKER' instead
```